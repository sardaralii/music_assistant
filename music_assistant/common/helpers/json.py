"""Helpers to work with (de)serializing of json."""

import asyncio
import base64
from _collections_abc import dict_keys, dict_values
from types import MethodType
from typing import Any, TypeVar

import aiofiles
import orjson
from mashumaro.mixins.orjson import DataClassORJSONMixin

JSON_ENCODE_EXCEPTIONS = (TypeError, ValueError)
JSON_DECODE_EXCEPTIONS = (orjson.JSONDecodeError,)

DO_NOT_SERIALIZE_TYPES = (MethodType, asyncio.Task)


def get_serializable_value(obj: Any, raise_unhandled: bool = False) -> Any:
    """Parse the value to its serializable equivalent."""
    if getattr(obj, "do_not_serialize", None):
        return None
    if (
        isinstance(obj, list | set | filter | tuple | dict_values | dict_keys | dict_values)
        or obj.__class__ == "dict_valueiterator"
    ):
        return [get_serializable_value(x) for x in obj]
    if hasattr(obj, "to_dict"):
        return obj.to_dict()
    if isinstance(obj, bytes):
        return base64.b64encode(obj).decode("ascii")
    if isinstance(obj, DO_NOT_SERIALIZE_TYPES):
        return None
    if raise_unhandled:
        raise TypeError
    return obj


def serialize_to_json(obj: Any) -> Any:
    """Serialize a value (or a list of values) to json."""
    if obj is None:
        return obj
    if hasattr(obj, "to_json"):
        return obj.to_json()
    return json_dumps(get_serializable_value(obj))


def json_dumps(data: Any, indent: bool = False) -> str:
    """Dump json string."""
    # we use the passthrough dataclass option because we use mashumaro for that
    option = orjson.OPT_OMIT_MICROSECONDS | orjson.OPT_PASSTHROUGH_DATACLASS
    if indent:
        option |= orjson.OPT_INDENT_2
    return orjson.dumps(
        data,
        default=get_serializable_value,
        option=option,
    ).decode("utf-8")


json_loads = orjson.loads

TargetT = TypeVar("TargetT", bound=DataClassORJSONMixin)


async def load_json_file(path: str, target_class: type[TargetT]) -> TargetT:
    """Load JSON from file."""
    async with aiofiles.open(path) as _file:
        content = await _file.read()
        return target_class.from_json(content)
