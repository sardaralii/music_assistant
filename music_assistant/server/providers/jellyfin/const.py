"""Constants for the Jellyfin integration."""

from typing import Final

from aiojellyfin import ImageType as JellyImageType
from aiojellyfin import ItemFields

from music_assistant.common.models.enums import ImageType, MediaType
from music_assistant.common.models.media_items import ItemMapping
from music_assistant.constants import UNKNOWN_ARTIST

DOMAIN: Final = "jellyfin"

CLIENT_VERSION: Final = "0.1"

COLLECTION_TYPE_MOVIES: Final = "movies"
COLLECTION_TYPE_MUSIC: Final = "music"
COLLECTION_TYPE_TVSHOWS: Final = "tvshows"

CONF_CLIENT_DEVICE_ID: Final = "client_device_id"

DEFAULT_NAME: Final = "Jellyfin"

ITEM_KEY_COLLECTION_TYPE: Final = "CollectionType"
ITEM_KEY_ID: Final = "Id"
ITEM_KEY_IMAGE_TAGS: Final = "ImageTags"
ITEM_KEY_INDEX_NUMBER: Final = "IndexNumber"
ITEM_KEY_MEDIA_SOURCES: Final = "MediaSources"
ITEM_KEY_MEDIA_TYPE: Final = "MediaType"
ITEM_KEY_MEDIA_STREAMS: Final = "MediaStreams"
ITEM_KEY_MEDIA_CHANNELS: Final = "Channels"
ITEM_KEY_MEDIA_CODEC: Final = "Codec"
ITEM_KEY_NAME: Final = "Name"
ITEM_KEY_PROVIDER_IDS: Final = "ProviderIds"
ITEM_KEY_PRODUCTION_YEAR: Final = "ProductionYear"
ITEM_KEY_OVERVIEW: Final = "Overview"
ITEM_KEY_MUSICBRAINZ_RELEASE_GROUP: Final = "MusicBrainzReleaseGroup"
ITEM_KEY_MUSICBRAINZ_ARTIST: Final = "MusicBrainzArtist"
ITEM_KEY_MUSICBRAINZ_ALBUM: Final = "MusicBrainzAlbum"
ITEM_KEY_MUSICBRAINZ_TRACK: Final = "MusicBrainzTrack"
ITEM_KEY_SORT_NAME: Final = "SortName"
ITEM_KEY_ALBUM_ARTIST: Final = "AlbumArtist"
ITEM_KEY_ALBUM_ARTISTS: Final = "AlbumArtists"
ITEM_KEY_ALBUM: Final = "Album"
ITEM_KEY_ALBUM_ID: Final = "AlbumId"
ITEM_KEY_ARTIST_ITEMS: Final = "ArtistItems"
ITEM_KEY_CAN_DOWNLOAD: Final = "CanDownload"
ITEM_KEY_PARENT_INDEX_NUM: Final = "ParentIndexNumber"
ITEM_KEY_RUNTIME_TICKS: Final = "RunTimeTicks"
ITEM_KEY_USER_DATA: Final = "UserData"

ITEM_TYPE_AUDIO: Final = "Audio"
ITEM_TYPE_LIBRARY: Final = "CollectionFolder"

USER_DATA_KEY_IS_FAVORITE: Final = "IsFavorite"

MAX_IMAGE_WIDTH: Final = 500
MAX_STREAMING_BITRATE: Final = "140000000"

MEDIA_SOURCE_KEY_PATH: Final = "Path"

MEDIA_TYPE_AUDIO: Final = "Audio"
MEDIA_TYPE_NONE: Final = ""

SUPPORTED_COLLECTION_TYPES: Final = [COLLECTION_TYPE_MUSIC]

SUPPORTED_CONTAINER_FORMATS: Final = "ogg,flac,mp3,aac,mpeg,alac,wav,aiff,wma,m4a,m4b,dsf,opus,wv"

PLAYABLE_ITEM_TYPES: Final = [ITEM_TYPE_AUDIO]

ARTIST_FIELDS: Final = [
    ItemFields.Overview,
    ItemFields.ProviderIds,
    ItemFields.SortName,
]
ALBUM_FIELDS: Final = [
    ItemFields.Overview,
    ItemFields.ProviderIds,
    ItemFields.SortName,
]
TRACK_FIELDS: Final = [
    ItemFields.ProviderIds,
    ItemFields.CanDownload,
    ItemFields.SortName,
    ItemFields.MediaSources,
    ItemFields.MediaStreams,
]

USER_APP_NAME: Final = "Music Assistant"
USER_AGENT: Final = "Music-Assistant-1.0"

UNKNOWN_ARTIST_MAPPING: Final = ItemMapping(
    media_type=MediaType.ARTIST, item_id=UNKNOWN_ARTIST, provider=DOMAIN, name=UNKNOWN_ARTIST
)

MEDIA_IMAGE_TYPES: Final = {
    JellyImageType.Primary: ImageType.THUMB,
    JellyImageType.Logo: ImageType.LOGO,
}
