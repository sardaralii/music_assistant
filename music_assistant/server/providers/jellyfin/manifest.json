{
  "type": "music",
  "domain": "jellyfin",
  "name": "Jellyfin Media Server Library",
  "description": "Support for the Jellyfin streaming provider in Music Assistant.",
  "codeowners": ["@lokiberra", "@Jc2k"],
  "requirements": ["aiojellyfin==0.10.1"],
  "documentation": "https://music-assistant.io/music-providers/jellyfin/",
  "multi_instance": true
}
