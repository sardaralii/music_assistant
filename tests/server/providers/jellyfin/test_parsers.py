"""Test we can parse Jellyfin models into Music Assistant models."""

import logging
import pathlib
from collections.abc import AsyncGenerator

import aiofiles
import aiohttp
import pytest
from aiojellyfin import Artist, Connection, SessionConfiguration
from mashumaro.codecs.json import JSONDecoder
from syrupy.assertion import SnapshotAssertion

from music_assistant.server.providers.jellyfin.parsers import parse_album, parse_artist, parse_track

FIXTURES_DIR = pathlib.Path(__file__).parent / "fixtures"
ARTIST_FIXTURES = list(FIXTURES_DIR.glob("artists/*.json"))
ALBUM_FIXTURES = list(FIXTURES_DIR.glob("albums/*.json"))
TRACK_FIXTURES = list(FIXTURES_DIR.glob("tracks/*.json"))

ARTIST_DECODER = JSONDecoder(Artist)

_LOGGER = logging.getLogger(__name__)


@pytest.fixture
async def connection() -> AsyncGenerator[Connection, None]:
    """Spin up a dummy connection."""
    async with aiohttp.ClientSession() as session:
        session_config = SessionConfiguration(
            session=session,
            url="http://localhost:1234",
            app_name="X",
            app_version="0.0.0",
            device_id="X",
            device_name="localhost",
        )
        yield Connection(session_config, "USER_ID", "ACCESS_TOKEN")


@pytest.mark.parametrize("example", ARTIST_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_artists(
    example: pathlib.Path, connection: Connection, snapshot: SnapshotAssertion
) -> None:
    """Test we can parse artists."""
    async with aiofiles.open(example) as fp:
        raw_data = ARTIST_DECODER.decode(await fp.read())
    parsed = parse_artist(_LOGGER, "xx-instance-id-xx", connection, raw_data).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed


@pytest.mark.parametrize("example", ALBUM_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_albums(
    example: pathlib.Path, connection: Connection, snapshot: SnapshotAssertion
) -> None:
    """Test we can parse albums."""
    async with aiofiles.open(example) as fp:
        raw_data = ARTIST_DECODER.decode(await fp.read())
    parsed = parse_album(_LOGGER, "xx-instance-id-xx", connection, raw_data).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed


@pytest.mark.parametrize("example", TRACK_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_tracks(
    example: pathlib.Path, connection: Connection, snapshot: SnapshotAssertion
) -> None:
    """Test we can parse tracks."""
    async with aiofiles.open(example) as fp:
        raw_data = ARTIST_DECODER.decode(await fp.read())
    parsed = parse_track(_LOGGER, "xx-instance-id-xx", connection, raw_data).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"]
    assert snapshot == parsed
