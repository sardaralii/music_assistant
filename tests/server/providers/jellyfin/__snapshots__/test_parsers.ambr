# serializer version: 1
# name: test_parse_albums[infest]
  dict({
    'album_type': 'unknown',
    'artists': list([
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'item_id': 'e439648e08ade14e27d5de48fa97c88e',
        'media_type': 'artist',
        'name': 'Papa Roach',
        'provider': 'xx-instance-id-xx',
        'sort_name': 'papa roach',
        'uri': 'xx-instance-id-xx://artist/e439648e08ade14e27d5de48fa97c88e',
        'version': '',
      }),
    ]),
    'external_ids': list([
      list([
        'musicbrainz_albumid',
        'bf25b030-0cbb-495a-8d79-6c7fee20a089',
      ]),
      list([
        'musicbrainz_releasegroupid',
        '0193355a-cdfb-3936-afd2-44d651eb006d',
      ]),
    ]),
    'favorite': False,
    'item_id': '70b7288088b42d318f75dbcc41fd0091',
    'media_type': 'album',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'http://localhost:1234/Items/70b7288088b42d318f75dbcc41fd0091/Images/Primary?api_key=ACCESS_TOKEN',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'review': None,
      'style': None,
    }),
    'name': 'Infest',
    'position': None,
    'provider': 'jellyfin',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 320,
          'channels': 2,
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '70b7288088b42d318f75dbcc41fd0091',
        'provider_domain': 'jellyfin',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': 'infest',
    'uri': 'jellyfin://album/70b7288088b42d318f75dbcc41fd0091',
    'version': '',
    'year': 2000,
  })
# ---
# name: test_parse_albums[this_is_christmas]
  dict({
    'album_type': 'unknown',
    'artists': list([
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'item_id': '555b36f7d310d1b7405557a8775c6878',
        'media_type': 'artist',
        'name': 'Emmy the Great & Tim Wheeler',
        'provider': 'xx-instance-id-xx',
        'sort_name': 'emmy the great & tim wheeler',
        'uri': 'xx-instance-id-xx://artist/555b36f7d310d1b7405557a8775c6878',
        'version': '',
      }),
    ]),
    'external_ids': list([
      list([
        'musicbrainz_albumid',
        'b13a174d-527d-44a1-b8f8-a4c78b03b7d9',
      ]),
      list([
        'musicbrainz_releasegroupid',
        'f002d6b7-17af-4f9e-8d30-5486548ffe6f',
      ]),
    ]),
    'favorite': False,
    'item_id': '32ed6a0091733dcff57eae67010f3d4b',
    'media_type': 'album',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'http://localhost:1234/Items/32ed6a0091733dcff57eae67010f3d4b/Images/Primary?api_key=ACCESS_TOKEN',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'review': None,
      'style': None,
    }),
    'name': 'This Is Christmas',
    'position': None,
    'provider': 'jellyfin',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 320,
          'channels': 2,
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '32ed6a0091733dcff57eae67010f3d4b',
        'provider_domain': 'jellyfin',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': 'this is christmas',
    'uri': 'jellyfin://album/32ed6a0091733dcff57eae67010f3d4b',
    'version': '',
    'year': 2011,
  })
# ---
# name: test_parse_albums[yesterday_when_i_was_mad]
  dict({
    'album_type': 'unknown',
    'artists': list([
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'item_id': '[unknown]',
        'media_type': 'artist',
        'name': '[unknown]',
        'provider': 'jellyfin',
        'sort_name': '[unknown]',
        'uri': 'jellyfin://artist/[unknown]',
        'version': '',
      }),
    ]),
    'external_ids': list([
    ]),
    'favorite': False,
    'item_id': '7c8d0bd55291c7fc0451d17ebef30017',
    'media_type': 'album',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
      ]),
      'label': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'review': None,
      'style': None,
    }),
    'name': 'Yesterday, When I Was Mad [Disc 2]',
    'position': None,
    'provider': 'jellyfin',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 320,
          'channels': 2,
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '7c8d0bd55291c7fc0451d17ebef30017',
        'provider_domain': 'jellyfin',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': 'yesterday when i was mad [disc 0000000002]',
    'uri': 'jellyfin://album/7c8d0bd55291c7fc0451d17ebef30017',
    'version': '',
    'year': None,
  })
# ---
# name: test_parse_artists[ash]
  dict({
    'external_ids': list([
      list([
        'musicbrainz_artistid',
        '99164692-c02d-407c-81c9-25d338dd21f4',
      ]),
    ]),
    'favorite': False,
    'item_id': 'dd954bbf54398e247d803186d3585b79',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'http://localhost:1234/Items/dd954bbf54398e247d803186d3585b79/Images/Backdrop/0?api_key=ACCESS_TOKEN',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'fanart',
        }),
        dict({
          'path': 'http://localhost:1234/Items/dd954bbf54398e247d803186d3585b79/Images/Primary?api_key=ACCESS_TOKEN',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
        dict({
          'path': 'http://localhost:1234/Items/dd954bbf54398e247d803186d3585b79/Images/Logo?api_key=ACCESS_TOKEN',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'logo',
        }),
      ]),
      'label': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'review': None,
      'style': None,
    }),
    'name': 'Ash',
    'position': None,
    'provider': 'jellyfin',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 320,
          'channels': 2,
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': 'dd954bbf54398e247d803186d3585b79',
        'provider_domain': 'jellyfin',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': 'ash',
    'uri': 'jellyfin://artist/dd954bbf54398e247d803186d3585b79',
    'version': '',
  })
# ---
# name: test_parse_tracks[thrown_away]
  dict({
    'album': dict({
      'available': True,
      'external_ids': list([
      ]),
      'image': None,
      'item_id': '70b7288088b42d318f75dbcc41fd0091',
      'media_type': 'album',
      'name': 'Unknown Album (70b7288088b42d318f75dbcc41fd0091)',
      'provider': 'xx-instance-id-xx',
      'sort_name': 'unknown album (70b7288088b42d318f75dbcc41fd0091)',
      'uri': 'xx-instance-id-xx://album/70b7288088b42d318f75dbcc41fd0091',
      'version': '',
    }),
    'artists': list([
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'item_id': '[unknown]',
        'media_type': 'artist',
        'name': '[unknown]',
        'provider': 'jellyfin',
        'sort_name': '[unknown]',
        'uri': 'jellyfin://artist/[unknown]',
        'version': '',
      }),
    ]),
    'disc_number': 0,
    'duration': 577,
    'external_ids': list([
    ]),
    'favorite': False,
    'item_id': 'b5319fb11cde39fca2023184fcfa9862',
    'media_type': 'track',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
      ]),
      'label': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'review': None,
      'style': None,
    }),
    'name': '11 Thrown Away',
    'position': 0,
    'provider': 'xx-instance-id-xx',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 320,
          'channels': 2,
          'content_type': 'mp3',
          'output_format_str': 'mp3',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': 'b5319fb11cde39fca2023184fcfa9862',
        'provider_domain': 'jellyfin',
        'provider_instance': 'xx-instance-id-xx',
        'url': 'http://localhost:1234/Audio/b5319fb11cde39fca2023184fcfa9862/universal?UserId=USER_ID&DeviceId=X&MaxStreamingBitrate=140000000&api_key=ACCESS_TOKEN',
      }),
    ]),
    'sort_name': '11 thrown away',
    'track_number': 0,
    'uri': 'xx-instance-id-xx://track/b5319fb11cde39fca2023184fcfa9862',
    'version': '',
  })
# ---
# name: test_parse_tracks[where_the_bands_are]
  dict({
    'album': None,
    'artists': list([
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'item_id': '94875b0dd58cbf5245a135982133651a',
        'media_type': 'artist',
        'name': 'Dead Like Harry',
        'provider': 'xx-instance-id-xx',
        'sort_name': 'dead like harry',
        'uri': 'xx-instance-id-xx://artist/94875b0dd58cbf5245a135982133651a',
        'version': '',
      }),
    ]),
    'disc_number': 1,
    'duration': 246,
    'external_ids': list([
    ]),
    'favorite': False,
    'item_id': '54918f75ee8f6c8b8dc5efd680644f29',
    'media_type': 'track',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'http://localhost:1234/Items/54918f75ee8f6c8b8dc5efd680644f29/Images/Primary?api_key=ACCESS_TOKEN',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'review': None,
      'style': None,
    }),
    'name': 'Where the Bands Are (2018 Version)',
    'position': 1,
    'provider': 'xx-instance-id-xx',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 320,
          'channels': 2,
          'content_type': 'aac',
          'output_format_str': 'aac',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '54918f75ee8f6c8b8dc5efd680644f29',
        'provider_domain': 'jellyfin',
        'provider_instance': 'xx-instance-id-xx',
        'url': 'http://localhost:1234/Audio/54918f75ee8f6c8b8dc5efd680644f29/universal?UserId=USER_ID&DeviceId=X&MaxStreamingBitrate=140000000&api_key=ACCESS_TOKEN',
      }),
    ]),
    'sort_name': 'where the bands are (2018 version)',
    'track_number': 1,
    'uri': 'xx-instance-id-xx://track/54918f75ee8f6c8b8dc5efd680644f29',
    'version': '',
  })
# ---
# name: test_parse_tracks[zombie_christmas]
  dict({
    'album': dict({
      'available': True,
      'external_ids': list([
      ]),
      'image': None,
      'item_id': '32ed6a0091733dcff57eae67010f3d4b',
      'media_type': 'album',
      'name': 'This Is Christmas',
      'provider': 'xx-instance-id-xx',
      'sort_name': 'this is christmas',
      'uri': 'xx-instance-id-xx://album/32ed6a0091733dcff57eae67010f3d4b',
      'version': '',
    }),
    'artists': list([
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'item_id': 'a0c459294295710546c81c20a8d9abfc',
        'media_type': 'artist',
        'name': 'Emmy the Great',
        'provider': 'xx-instance-id-xx',
        'sort_name': 'emmy the great',
        'uri': 'xx-instance-id-xx://artist/a0c459294295710546c81c20a8d9abfc',
        'version': '',
      }),
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'item_id': '1952db245ddef4e41dcd016475379190',
        'media_type': 'artist',
        'name': 'Tim Wheeler',
        'provider': 'xx-instance-id-xx',
        'sort_name': 'tim wheeler',
        'uri': 'xx-instance-id-xx://artist/1952db245ddef4e41dcd016475379190',
        'version': '',
      }),
    ]),
    'disc_number': 1,
    'duration': 224,
    'external_ids': list([
      list([
        'musicbrainz_recordingid',
        '17d1019d-d4f4-326c-b4bb-d8aec2607bd7',
      ]),
    ]),
    'favorite': False,
    'item_id': 'fb12a77f49616a9fc56a6fab2b94174c',
    'media_type': 'track',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'http://localhost:1234/Items/fb12a77f49616a9fc56a6fab2b94174c/Images/Primary?api_key=ACCESS_TOKEN',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'review': None,
      'style': None,
    }),
    'name': 'Zombie Christmas',
    'position': 8,
    'provider': 'xx-instance-id-xx',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 320,
          'channels': 2,
          'content_type': 'aac',
          'output_format_str': 'aac',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': 'fb12a77f49616a9fc56a6fab2b94174c',
        'provider_domain': 'jellyfin',
        'provider_instance': 'xx-instance-id-xx',
        'url': 'http://localhost:1234/Audio/fb12a77f49616a9fc56a6fab2b94174c/universal?UserId=USER_ID&DeviceId=X&MaxStreamingBitrate=140000000&api_key=ACCESS_TOKEN',
      }),
    ]),
    'sort_name': 'zombie christmas',
    'track_number': 8,
    'uri': 'xx-instance-id-xx://track/fb12a77f49616a9fc56a6fab2b94174c',
    'version': '',
  })
# ---
